File: java-backend/src/main/java/com/study/service/NoteService.java
package com.study.service;

import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.study.domain.Note;
import com.study.domain.NoteDocumentState;
import com.study.domain.NoteMember;
import com.study.domain.User;
import com.study.exception.ForbiddenException;
import com.study.exception.ResourceNotFoundException;
import com.study.mapper.NoteDocumentMapper;
import com.study.mapper.NoteMapper;
import com.study.mapper.NoteMemberMapper;
import com.study.mapper.UserMapper;
import com.study.security.JwtService;
import com.study.web.dto.CollabTokenResponse;
import com.study.web.dto.NoteMemberResponse;
import com.study.web.dto.NoteRequest;
import com.study.web.dto.NoteResponse;
import java.util.Base64;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/** 实时协作笔记业务:元数据、成员权限、协作 token、Yjs 快照持久化。 */
@Service
public class NoteService {

    private static final long COLLAB_TOKEN_TTL_SECONDS = 300L;

    private final NoteMapper noteMapper;
    private final NoteMemberMapper noteMemberMapper;
    private final NoteDocumentMapper noteDocumentMapper;
    private final UserMapper userMapper;
    private final JwtService jwtService;
    private final String collabUrl;

    public NoteService(NoteMapper noteMapper,
                       NoteMemberMapper noteMemberMapper,
                       NoteDocumentMapper noteDocumentMapper,
                       UserMapper userMapper,
                       JwtService jwtService,
                       @Value("${collab.url:ws://localhost:19082}") String collabUrl) {
        this.noteMapper = noteMapper;
        this.noteMemberMapper = noteMemberMapper;
        this.noteDocumentMapper = noteDocumentMapper;
        this.userMapper = userMapper;
        this.jwtService = jwtService;
        this.collabUrl = collabUrl;
    }

    @Transactional(readOnly = true)
    public List<NoteResponse> list(Long userId) {
        List<NoteMember> memberships = noteMemberMapper.selectList(new LambdaQueryWrapper<NoteMember>()
                .eq(NoteMember::getUserId, userId)
                .orderByDesc(NoteMember::getCreatedAt));
        if (memberships.isEmpty()) {
            return List.of();
        }
        Map<Long, String> roles = memberships.stream()
                .collect(Collectors.toMap(NoteMember::getNoteId, NoteMember::getRole, (a, b) -> a));
        List<Long> noteIds = memberships.stream().map(NoteMember::getNoteId).toList();
        return noteMapper.selectBatchIds(noteIds).stream()
                .map(note -> toResponse(note, roles.get(note.getId())))
                .toList();
    }

    @Transactional(readOnly = true)
    public NoteResponse get(Long id, Long userId) {
        Note note = requireNote(id);
        String role = requireRole(id, userId);
        return toResponse(note, role);
    }

    @Transactional
    public NoteResponse create(NoteRequest request, Long ownerId) {
        Note note = new Note();
        note.setTitle(request.title());
        note.setOwnerId(ownerId);
        noteMapper.insert(note);

        NoteMember owner = new NoteMember();
        owner.setNoteId(note.getId());
        owner.setUserId(ownerId);
        owner.setRole("OWNER");
        noteMemberMapper.insert(owner);
        noteDocumentMapper.insertEmpty(note.getId(), new byte[0]);
        return toResponse(note, "OWNER");
    }

    @Transactional
    public NoteResponse update(Long id, NoteRequest request, Long userId) {
        Note note = requireNote(id);
        requireOwner(note, userId);
        note.setTitle(request.title());
        noteMapper.updateById(note);
        return toResponse(note, "OWNER");
    }

    @Transactional
    public void delete(Long id, Long userId) {
        Note note = requireNote(id);
        requireOwner(note, userId);
        noteMapper.deleteById(id);
    }

    @Transactional(readOnly = true)
    public List<NoteMemberResponse> members(Long noteId, Long userId) {
        requireRole(noteId, userId);
        List<NoteMember> members = noteMemberMapper.selectList(new LambdaQueryWrapper<NoteMember>()
                .eq(NoteMember::getNoteId, noteId));
        List<Long> userIds = members.stream().map(NoteMember::getUserId).toList();
        Map<Long, String> usernames = userMapper.selectBatchIds(userIds).stream()
                .collect(Collectors.toMap(User::getId, User::getUsername));
        return members.stream()
                .map(member -> new NoteMemberResponse(
                        member.getUserId(), usernames.get(member.getUserId()), member.getRole()))
                .toList();
    }

    @Transactional
    public NoteMemberResponse addMember(Long noteId, Long ownerId, Long memberUserId, String role) {
        Note note = requireNote(noteId);
        requireOwner(note, ownerId);
        User user = userMapper.selectById(memberUserId);
        if (user == null) {
            throw new ResourceNotFoundException("用户不存在");
        }
        NoteMember existing = findMember(noteId, memberUserId);
        if (existing == null) {
            NoteMember member = new NoteMember();
            member.setNoteId(noteId);
            member.setUserId(memberUserId);
            member.setRole(role == null ? "EDITOR" : role);
            noteMemberMapper.insert(member);
            return new NoteMemberResponse(memberUserId, user.getUsername(), member.getRole());
        }
        if ("OWNER".equals(existing.getRole())) {
            throw new ForbiddenException("不能修改笔记拥有者角色");
        }
        existing.setRole(role == null ? "EDITOR" : role);
        noteMemberMapper.updateById(existing);
        return new NoteMemberResponse(memberUserId, user.getUsername(), existing.getRole());
    }

    @Transactional
    public void removeMember(Long noteId, Long ownerId, Long memberUserId) {
        Note note = requireNote(noteId);
        requireOwner(note, ownerId);
        if (note.getOwnerId().equals(memberUserId)) {
            throw new ForbiddenException("不能移除笔记拥有者");
        }
        NoteMember member = findMember(noteId, memberUserId);
        if (member != null) {
            noteMemberMapper.deleteById(member.getId());
        }
    }

    @Transactional(readOnly = true)
    public CollabTokenResponse collabToken(Long noteId, Long userId, String username) {
        String role = requireRole(noteId, userId);
        if ("VIEWER".equals(role)) {
            throw new ForbiddenException("只读成员暂不能进入协作编辑");
        }
        String token = jwtService.generateCollabToken(userId, username, noteId, role, COLLAB_TOKEN_TTL_SECONDS);
        return new CollabTokenResponse(token, COLLAB_TOKEN_TTL_SECONDS, "note:" + noteId, role, collabUrl);
    }

    @Transactional(readOnly = true)
    public String loadDocumentState(Long noteId) {
        requireNote(noteId);
        NoteDocumentState document = noteDocumentMapper.selectState(noteId);
        byte[] state = document == null ? null : document.getYdocState();
        if (state == null || state.length == 0) {
            return "";
        }
        return Base64.getEncoder().encodeToString(state);
    }

    @Transactional
    public void saveDocumentState(Long noteId, String base64State) {
        requireNote(noteId);
        byte[] state = base64State == null || base64State.isBlank()
                ? new byte[0]
                : Base64.getDecoder().decode(base64State);
        int updated = noteDocumentMapper.updateState(noteId, state);
        if (updated == 0) {
            noteDocumentMapper.insertEmpty(noteId, state);
        }
    }

    private Note requireNote(Long id) {
        Note note = noteMapper.selectById(id);
        if (note == null) {
            throw new ResourceNotFoundException("笔记不存在");
        }
        return note;
    }

    private String requireRole(Long noteId, Long userId) {
        NoteMember member = findMember(noteId, userId);
        if (member == null) {
            throw new ForbiddenException("无权访问该笔记");
        }
        return member.getRole();
    }

    private void requireOwner(Note note, Long userId) {
        if (!note.getOwnerId().equals(userId)) {
            throw new ForbiddenException("只有笔记拥有者可以操作");
        }
    }

    private NoteMember findMember(Long noteId, Long userId) {
        return noteMemberMapper.selectOne(new LambdaQueryWrapper<NoteMember>()
                .eq(NoteMember::getNoteId, noteId)
                .eq(NoteMember::getUserId, userId));
    }

    private NoteResponse toResponse(Note note, String role) {
        User owner = userMapper.selectById(note.getOwnerId());
        String ownerUsername = owner == null ? null : owner.getUsername();
        return new NoteResponse(
                note.getId(), note.getTitle(), note.getOwnerId(), ownerUsername,
                role, note.getCreatedAt(), note.getUpdatedAt());
    }
}

File: kotlin-backend/src/main/kotlin/com/study/service/NoteService.kt
package com.study.service

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper
import com.study.domain.Note
import com.study.domain.NoteMember
import com.study.exception.ForbiddenException
import com.study.exception.ResourceNotFoundException
import com.study.mapper.NoteDocumentMapper
import com.study.mapper.NoteMapper
import com.study.mapper.NoteMemberMapper
import com.study.mapper.UserMapper
import com.study.security.JwtService
import com.study.web.dto.CollabTokenResponse
import com.study.web.dto.NoteMemberResponse
import com.study.web.dto.NoteRequest
import com.study.web.dto.NoteResponse
import java.util.Base64
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * 实时协作笔记业务:元数据、成员权限、协作 token、Yjs 快照持久化。
 *
 * 设计重点:
 * - Spring 后端是权限和数据的事实源,负责 note/note_members/note_documents 三张表;
 * - Nest/Hocuspocus 只负责实时同步,每次连接前必须拿这里签发的短期 collab token;
 * - Yjs 文档是二进制 update,数据库用 BLOB,接口层用 Base64 包一层。
 */
// @Service: 让 NoteService 成为可注入的业务 Bean;Controller 不手动 new,由 Spring 容器装配。
@Service
class NoteService(
    private val noteMapper: NoteMapper,
    private val noteMemberMapper: NoteMemberMapper,
    private val noteDocumentMapper: NoteDocumentMapper,
    private val userMapper: UserMapper,
    private val jwtService: JwtService,
    // @Value: 从配置读取协作服务地址,用于返回给前端连接 Hocuspocus。
    @Value("\${collab.url:ws://localhost:19082/collab/notes}") private val collabUrl: String,
) {
    companion object {
        private const val COLLAB_TOKEN_TTL_SECONDS = 300L
    }

    // @Transactional(readOnly = true): 只读事务,用于成员/笔记查询和权限校验。
    @Transactional(readOnly = true)
    fun list(userId: Long): List<NoteResponse> {
        val memberships = noteMemberMapper.selectList(
            QueryWrapper<NoteMember>()
                .eq("user_id", userId)
                .orderByDesc("created_at"),
        )
        if (memberships.isEmpty()) return emptyList()
        val roles = memberships.associate { it.noteId!! to it.role!! }
        val noteIds = memberships.mapNotNull { it.noteId }
        return noteMapper.selectBatchIds(noteIds).map { toResponse(it, roles[it.id] ?: "VIEWER") }
    }

    @Transactional(readOnly = true)
    fun get(id: Long, userId: Long): NoteResponse {
        val note = requireNote(id)
        val role = requireRole(id, userId)
        return toResponse(note, role)
    }

    // @Transactional: 创建笔记、OWNER 成员、空文档快照必须在一个事务内成功或回滚。
    @Transactional
    fun create(request: NoteRequest, ownerId: Long): NoteResponse {
        // 创建笔记时同时创建 OWNER 成员和空文档快照,保证后续协作服务一定能 load 到记录。
        val note = Note().apply {
            title = request.title
            this.ownerId = ownerId
        }
        noteMapper.insert(note)

        val owner = NoteMember().apply {
            noteId = note.id
            userId = ownerId
            role = "OWNER"
        }
        noteMemberMapper.insert(owner)
        noteDocumentMapper.insertEmpty(note.id!!, ByteArray(0))
        return toResponse(note, "OWNER")
    }

    @Transactional
    fun update(id: Long, request: NoteRequest, userId: Long): NoteResponse {
        val note = requireNote(id)
        requireOwner(note, userId)
        note.title = request.title
        noteMapper.updateById(note)
        return toResponse(note, "OWNER")
    }

    @Transactional
    fun delete(id: Long, userId: Long) {
        val note = requireNote(id)
        requireOwner(note, userId)
        noteMapper.deleteById(id)
    }

    @Transactional(readOnly = true)
    fun members(noteId: Long, userId: Long): List<NoteMemberResponse> {
        requireRole(noteId, userId)
        val members = noteMemberMapper.selectList(QueryWrapper<NoteMember>().eq("note_id", noteId))
        val userIds = members.mapNotNull { it.userId }
        val usernames = userMapper.selectBatchIds(userIds).associate { it.id!! to it.username }
        return members.map { NoteMemberResponse(it.userId, usernames[it.userId], it.role) }
    }

    @Transactional
    fun addMember(noteId: Long, ownerId: Long, memberUserId: Long, role: String?): NoteMemberResponse {
        val note = requireNote(noteId)
        requireOwner(note, ownerId)
        val user = userMapper.selectById(memberUserId) ?: throw ResourceNotFoundException("用户不存在")
        val existing = findMember(noteId, memberUserId)
        if (existing == null) {
            val member = NoteMember().apply {
                this.noteId = noteId
                userId = memberUserId
                this.role = role ?: "EDITOR"
            }
            noteMemberMapper.insert(member)
            return NoteMemberResponse(memberUserId, user.username, member.role)
        }
        if (existing.role == "OWNER") {
            throw ForbiddenException("不能修改笔记拥有者角色")
        }
        existing.role = role ?: "EDITOR"
        noteMemberMapper.updateById(existing)
        return NoteMemberResponse(memberUserId, user.username, existing.role)
    }

    @Transactional
    fun removeMember(noteId: Long, ownerId: Long, memberUserId: Long) {
        val note = requireNote(noteId)
        requireOwner(note, ownerId)
        if (note.ownerId == memberUserId) {
            throw ForbiddenException("不能移除笔记拥有者")
        }
        findMember(noteId, memberUserId)?.let { noteMemberMapper.deleteById(it.id) }
    }

    @Transactional(readOnly = true)
    fun collabToken(noteId: Long, userId: Long, username: String): CollabTokenResponse {
        // 普通登录 JWT 只给业务 API 使用;协作 WS 只认 typ=collab 的短期 token,降低泄漏后的影响面。
        val role = requireRole(noteId, userId)
        if (role == "VIEWER") {
            throw ForbiddenException("只读成员暂不能进入协作编辑")
        }
        val token = jwtService.generateCollabToken(userId, username, noteId, role, COLLAB_TOKEN_TTL_SECONDS)
        return CollabTokenResponse(token, COLLAB_TOKEN_TTL_SECONDS, "note:$noteId", role, collabUrl)
    }

    @Transactional(readOnly = true)
    fun loadDocumentState(noteId: Long): String {
        // 内部接口也先确认笔记存在,避免协作服务创建不存在的野文档。
        requireNote(noteId)
        val state = noteDocumentMapper.selectState(noteId)?.ydocState
        return if (state == null || state.isEmpty()) "" else Base64.getEncoder().encodeToString(state)
    }

    @Transactional
    fun saveDocumentState(noteId: Long, base64State: String?) {
        // Hocuspocus 存的是 Yjs update 二进制;HTTP JSON 层传 Base64,到库前转回 ByteArray。
        requireNote(noteId)
        val state = if (base64State.isNullOrBlank()) ByteArray(0) else Base64.getDecoder().decode(base64State)
        val updated = noteDocumentMapper.updateState(noteId, state)
        if (updated == 0) {
            noteDocumentMapper.insertEmpty(noteId, state)
        }
    }

    private fun requireNote(id: Long): Note =
        noteMapper.selectById(id) ?: throw ResourceNotFoundException("笔记不存在")

    private fun requireRole(noteId: Long, userId: Long): String =
        findMember(noteId, userId)?.role ?: throw ForbiddenException("无权访问该笔记")

    private fun requireOwner(note: Note, userId: Long) {
        if (note.ownerId != userId) {
            throw ForbiddenException("只有笔记拥有者可以操作")
        }
    }

    private fun findMember(noteId: Long, userId: Long): NoteMember? =
        noteMemberMapper.selectOne(
            QueryWrapper<NoteMember>()
                .eq("note_id", noteId)
                .eq("user_id", userId),
        )

    private fun toResponse(note: Note, role: String): NoteResponse {
        val owner = note.ownerId?.let { userMapper.selectById(it) }
        return NoteResponse(note.id, note.title, note.ownerId, owner?.username, role, note.createdAt, note.updatedAt)
    }
}
